File: 2025/3/joltage.rb
#!/usr/bin/env ruby

def main
  f = open("input.txt")
  sum = 0
  f.each_line do |line|
    chars = line.chomp.chars
    digits = []

    (0..11).each do |i|
      digit = if i == 11
        chars.max
      else
        chars[...-(11-i)].max
      end

      digits.push(digit)
      chars = chars[chars.index(digit) + 1..]
    end

    s = digits.join
    p "#{s}"
    sum += s.to_i
  end

  p "Sum: #{sum}"
end

main()

File: 2025/3/joltage.rs
fn main() {
	use std::fs::File;
	use std::io::{BufRead, BufReader};

	let mut sum: i64 = 0;

	let f = File::open("input.txt").unwrap();
	let reader = BufReader::new(f);

	for line in reader.lines() {
		let s = line.unwrap();

		let length = s.len();
		let mut left = 0;
		let mut digits = String::new();

		for i in 0..=11 {
			let mut max = '\0';
			let mut maxj = 0;
			let right = length - (11 - i);
			let slice = &s[left..right];
			for (j, c) in slice.chars().enumerate() {
				if c > max {
					max = c;
					maxj = left + j;
				}
			}
			digits.push(max);
			left = maxj + 1;
		}
		println!("{}", digits);
		sum += digits.parse::<i64>().unwrap();
	}

	println!("Sum: {}", sum);
}
